alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs

```diff
@@ -0,0 +1,438 @@
+            )
+
+{-# COMPLETE TxBody #-}
+
+instance Era era => HashAnnotated (TxBody era) era where
+  type HashIndex (TxBody era) = EraIndependentTxBody
+
+--------------------------------------------------------------------------------
+-- Serialisation
+--------------------------------------------------------------------------------
+
+instance Era era => ToCBOR (TxIn era) where
+  toCBOR (TxInCompact txId idx isFee) =
+    encode $
+      Rec TxInCompact
+        !> To txId
+        !> To idx
+        !> To isFee
+
+instance Era era => FromCBOR (TxIn era) where
+  fromCBOR = decode $ RecD TxInCompact <! From <! From <! From
+
+instance
+  (Era era, ToCBOR (CompactForm (Core.Value era))) =>
+  ToCBOR (TxOut era)
+  where
+  toCBOR (TxOutCompact addr cv dh) =
+    encode $
+      Rec TxOutCompact !> To addr !> To cv !> To dh
+
+instance
+  ( Era era,
+    DecodeNonNegative (CompactForm (Core.Value era)),
+    Compactible (Core.Value era)
+  ) =>
+  FromCBOR (TxOut era)
+  where
+  fromCBOR =
+    decode $
+      RecD TxOutCompact
+        <! From
+        <! D decodeNonNegative
+        <! From
+
+encodeTxBodyRaw ::
+  ( Era era,
+    Typeable (Core.Metadata era),
+    Typeable (Core.Script era),
+    ToCBOR (CompactForm (Core.Value era)),
+    ToCBOR (Core.Script era)
+  ) =>
+  TxBodyRaw era ->
+  Encode ('Closed 'Sparse) (TxBodyRaw era)
+encodeTxBodyRaw
+  TxBodyRaw
+    { _inputs,
+      _outputs,
+      _certs,
+      _wdrls,
+      _txfee,
+      _vldt = ValidityInterval bot top,
+      _update,
+      _mdHash,
+      _mint,
+      _exunits,
+      _scriptHash
+    } =
+    Keyed
+      ( \i o f t c w u mh b mi e s ->
+          TxBodyRaw i o c w f (ValidityInterval b t) u mh mi e s
+      )
+      !> Key 0 (E encodeFoldable _inputs)
+      !> Key 1 (E encodeFoldable _outputs)
+      !> Key 2 (To _txfee)
+      !> encodeKeyedStrictMaybe 3 top
+      !> Omit null (Key 4 (E encodeFoldable _certs))
+      !> Omit (null . unWdrl) (Key 5 (To _wdrls))
+      !> encodeKeyedStrictMaybe 6 _update
+      !> encodeKeyedStrictMaybe 7 _mdHash
+      !> encodeKeyedStrictMaybe 8 bot
+      !> Omit isZero (Key 9 (E encodeMint _mint))
+      !> Omit (== mempty) (Key 10 (To _exunits))
+      !> encodeKeyedStrictMaybe 11 _scriptHash
+    where
+      encodeKeyedStrictMaybe key x =
+        Omit isSNothing (Key key (E (toCBOR . fromSJust) x))
+
+      isSNothing :: StrictMaybe a -> Bool
+      isSNothing SNothing = True
+      isSNothing _ = False
+
+      fromSJust :: StrictMaybe a -> a
+      fromSJust (SJust x) = x
+      fromSJust SNothing = error "SNothing in fromSJust"
+
+instance
+  ( Era era,
+    Typeable (Core.Script era),
+    Typeable (Core.Metadata era),
+    Val (Core.Value era),
+    Compactible (Core.Value era),
+    Show (Core.Value era),
+    DecodeNonNegative (Core.Value era),
+    FromCBOR (Annotator (Core.Script era))
+  ) =>
+  FromCBOR (TxBodyRaw era)
+  where
+  fromCBOR = decode $ SparseKeyed "TxBodyRaw" init bodyFields []
```

Comment:
The final argument to `SparseKeyed` here is the set of required fields, such that an error will be raised if they're not present. So should set it here to the same set as in shelley-ma